File: biometric/src/main/java/dev/skomlach/biometric/compat/utils/SettingsHelper.java
package dev.skomlach.biometric.compat.utils;

import android.content.Context;
import android.provider.Settings;

import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;

@RestrictTo(RestrictTo.Scope.LIBRARY)
public class SettingsHelper {
    public static int getInt(Context context, String secureSettingKey, int defaultValue) {
        return (int) getLong(context, secureSettingKey, defaultValue);
    }

    public static long getLong(Context context, String secureSettingKey, long defaultValue) {
        long result = getLongInternal(context, secureSettingKey, defaultValue);
        if (result == defaultValue) {
            result = getIntInternal(context, secureSettingKey, (int) defaultValue);
        }
        return result;
    }

    public static String getString(Context context, String secureSettingKey, @NonNull String defaultValue) {

        try {
            String result = Settings.Secure.getString(context.getContentResolver(), secureSettingKey);

            if (!defaultValue.equals(result))
                return result;
        } catch (Throwable e) {

        }
        //fallback
        try {
            String result = Settings.System.getString(context.getContentResolver(), secureSettingKey);

            if (!defaultValue.equals(result))
                return result;
        } catch (Throwable e) {

        }
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.JELLY_BEAN_MR1)
            try {
                String result = Settings.Global.getString(context.getContentResolver(), secureSettingKey);

                if (!defaultValue.equals(result))
                    return result;
            } catch (Throwable e) {

            }
        return defaultValue;
    }

    private static long getLongInternal(Context context, String secureSettingKey, long defaultValue) {
        try {
            long result = Settings.Secure.getLong(context.getContentResolver(), secureSettingKey);

            if (result != defaultValue)
                return result;
        } catch (Throwable e) {

        }
        //fallback
        try {
            long result = Settings.System.getLong(context.getContentResolver(), secureSettingKey);

            if (result != defaultValue)
                return result;
        } catch (Throwable e) {

        }
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.JELLY_BEAN_MR1)
            try {
                long result = Settings.Global.getLong(context.getContentResolver(), secureSettingKey);

                if (result != defaultValue)
                    return result;
            } catch (Throwable e) {

            }
        return defaultValue;
    }

    private static int getIntInternal(Context context, String secureSettingKey, int defaultValue) {
        try {
            int result = Settings.Secure.getInt(context.getContentResolver(), secureSettingKey);

            if (result != defaultValue)
                return result;
        } catch (Throwable e) {

        }
        //fallback
        try {
            int result = Settings.System.getInt(context.getContentResolver(), secureSettingKey);

            if (result != defaultValue)
                return result;
        } catch (Throwable e) {

        }
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.JELLY_BEAN_MR1)
            try {
                int result = Settings.Global.getInt(context.getContentResolver(), secureSettingKey);

                if (result != defaultValue)
                    return result;
            } catch (Throwable e) {

            }
        return defaultValue;
    }
}

File: biometric/src/main/java/dev/skomlach/biometric/compat/utils/SettingsHelper.kt
package dev.skomlach.biometric.compat.utils

import android.content.Context
import android.os.Build
import android.provider.Settings
import androidx.annotation.RestrictTo

@RestrictTo(RestrictTo.Scope.LIBRARY)
object SettingsHelper {
    @JvmStatic
    fun getInt(context: Context, secureSettingKey: String?, defaultValue: Int): Int {
        return getLong(context, secureSettingKey, defaultValue.toLong()).toInt()
    }
    @JvmStatic
    fun getLong(context: Context, secureSettingKey: String?, defaultValue: Long): Long {
        var result = getLongInternal(context, secureSettingKey, defaultValue)
        if (result == defaultValue) {
            result = getIntInternal(context, secureSettingKey, defaultValue.toInt()).toLong()
        }
        return result
    }
    @JvmStatic
    fun getString(context: Context, secureSettingKey: String?, defaultValue: String): String {
        try {
            val result = Settings.Secure.getString(context.contentResolver, secureSettingKey)
            if (defaultValue != result) return result
        } catch (e: Throwable) {
        }
        //fallback
        try {
            val result = Settings.System.getString(context.contentResolver, secureSettingKey)
            if (defaultValue != result) return result
        } catch (e: Throwable) {
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) try {
            val result = Settings.Global.getString(context.contentResolver, secureSettingKey)
            if (defaultValue != result) return result
        } catch (e: Throwable) {
        }
        return defaultValue
    }

    private fun getLongInternal(
        context: Context,
        secureSettingKey: String?,
        defaultValue: Long
    ): Long {
        try {
            val result = Settings.Secure.getLong(context.contentResolver, secureSettingKey)
            if (result != defaultValue) return result
        } catch (e: Throwable) {
        }
        //fallback
        try {
            val result = Settings.System.getLong(context.contentResolver, secureSettingKey)
            if (result != defaultValue) return result
        } catch (e: Throwable) {
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) try {
            val result = Settings.Global.getLong(context.contentResolver, secureSettingKey)
            if (result != defaultValue) return result
        } catch (e: Throwable) {
        }
        return defaultValue
    }

    private fun getIntInternal(
        context: Context,
        secureSettingKey: String?,
        defaultValue: Int
    ): Int {
        try {
            val result = Settings.Secure.getInt(context.contentResolver, secureSettingKey)
            if (result != defaultValue) return result
        } catch (e: Throwable) {
        }
        //fallback
        try {
            val result = Settings.System.getInt(context.contentResolver, secureSettingKey)
            if (result != defaultValue) return result
        } catch (e: Throwable) {
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) try {
            val result = Settings.Global.getInt(context.contentResolver, secureSettingKey)
            if (result != defaultValue) return result
        } catch (e: Throwable) {
        }
        return defaultValue
    }
}